[CI] add qt action to all platform

terrychan999 opened the pull request:
In this PR, we manage Qt installation for all platforms using `jurplel/install-qt-action@v3`.
All the Qt versions are fixed to 6.5.3 LTS.
Also, `jurplel/install-qt-action@v3` has built-in `actions/setup-python@v4`, so we can avoid `brew unlink && brew link` issues like those in https://github.com/solvcon/modmesh/issues/289.

terrychan999 commented:
I just notice that https://github.com/solvcon/modmesh/issues/289 still exist in `lint / tidy_flake8_macos` because llvm@16 depends on python 3.12 which install from homebrew...

yungyuc commented:
Is this PR intended for discussion or review?  When filing a PR, please write explicit words for the intention.

terrychan999 committed:
add qt action, pin the version to 6.5.3

terrychan999 commented:
@yungyuc this PR is now ready for review, thanks.

yungyuc commented on .github/workflows/devbuild.yml:
```diff
@@ -113,11 +113,22 @@ jobs:
-              qt6-3d-dev xvfb \
-              libgl1-mesa-dev
-        fi
+        # if [ "${{ matrix.os }}" == "ubuntu-22.04" ] ; then \
```
Please add comments to explain why the block of commands is commented out.

yungyuc replied:
I don't like the style of comments:

```
# # ...
```

It unnecessarily makes the lines longer.

And please write "... cmake won't find the correct ...".  It seems "find out" isn't what you meant.

yungyuc commented on .github/workflows/devbuild.yml:
```diff
@@ -258,7 +269,18 @@ jobs:
             brew list -1 | grep python | while read formula; do brew unlink $formula; brew link --overwrite $formula; done;
           fi
-          brew install qt6
+          # brew install qt6
```
Please add comments to explain why the command is commented out.

yungyuc commented on .github/workflows/lint.yml:
```diff
@@ -78,11 +78,22 @@ jobs:
-              qt6-3d-dev xvfb \
-              libgl1-mesa-dev
-        fi
+        # if [ "${{ matrix.os }}" == "ubuntu-22.04" ] ; then \
```
Please add comments to explain why the block of commands is commented out.

yungyuc reviewed (changes requested):
Thanks, @terrychan999 .  Could you please address the following point?

- [x] Inline comment to explain the reason to disable code.

tigercosmos commented on .github/workflows/devbuild.yml:
```diff
@@ -258,7 +269,18 @@ jobs:
+      - name: Install Qt
+        uses: jurplel/install-qt-action@v3
+        with:
+          aqtversion: '==3.1.*'
```
`==3.1` should be fine?

terrychan999 replied:
this is a version specifier using in `pip install aqtinstall==<version specifier>`
https://github.com/jurplel/install-qt-action/blob/master/action/src/main.ts#L354

the version is just fellowing the current default.
https://github.com/jurplel/install-qt-action/tree/master?tab=readme-ov-file#aqtversion

tigercosmos replied:
True, just realized it follows pip's rule.
Thanks for the explanation.

terrychan999 committed:
add reason for disabling to legacy qt install method

terrychan999 commented on .github/workflows/devbuild.yml:
```diff
@@ -107,17 +107,31 @@ jobs:
+        #       libgl1-mesa-dev
+        # fi
+
+        # # This has been replaced by the 'Install Qt' section below to manage qt6 versioning independently from the OS.
```
Added explanation for disabling here.

yungyuc replied:
Explanation should go before the commented-out code.

terrychan999 commented on .github/workflows/devbuild.yml:
```diff
@@ -258,7 +272,20 @@ jobs:
-          brew install qt6
+
+          # brew install qt6
+          # # This has been replaced by the 'Install Qt' section below to manage qt6 versioning independently from the OS.
```
Added explanation for disabling here.

yungyuc replied:
Please move the explanation above the code commented out.  Please avoid doubling the commenting mark like `# # ...`

terrychan999 commented on .github/workflows/lint.yml:
```diff
@@ -71,18 +71,32 @@ jobs:
+        #       libgl1-mesa-dev
+        # fi
+
+        # # This has been replaced by the 'Install Qt' section below to manage qt6 versioning independently from the OS.
```
Added explanation for disabling here.

terrychan999 commented on .github/workflows/lint.yml:
```diff
@@ -171,10 +185,25 @@ jobs:
-          brew install llvm@16 qt6
+
+          # brew install llvm@16 qt6
+          # # This has been replaced by the 'Install Qt' section below to manage qt6 versioning independently from the OS.
```
Added explanation for disabling here.

yungyuc reviewed (changes requested):
The comments are not placed in right places.  Please review and update them all.

yungyuc commented:
@terrychan999 any comments?

terrychan999 committed:
better comments style and placement

terrychan999 committed:
disabling PEP668 with ENVVAR, change qt ver to 6.6.3

terrychan999 commented:
I've moved the comments to the position before the code and also removed the usage of double commenting mark.

I also added two changes:
1. Adding `PIP_BREAK_SYSTEM_PACKAGES: 1` in the env section of macOS-CI to avoid an error like the one in https://github.com/solvcon/modmesh/pull/301 with `jurplel/install-qt-action@v3`.
I used this environment variable because I couldn't directly pass the argument `--break-system-packages` into the action.

2. Changing the Qt version to `6.6.3`.
The reason for updating the Qt version to `6.6.3` is that `6.5.3 LTS` does not support Python 3.12, here's the reference:
https://wiki.qt.io/Qt_for_Python#Python_compatibility_matrix

yungyuc commented on .github/workflows/devbuild.yml:
```diff
@@ -112,12 +112,25 @@ jobs:
-              libgl1-mesa-dev
-        fi
+        # installing this package, cmake won't find the correct lib path.
+        # This has been replaced by the 'Install Qt' section below to manage qt6 versioning independently from the OS.
```
You didn't pay attention to the line width (79-char).  The overly long lines make it harder to maintain.

yungyuc commented on .github/workflows/devbuild.yml:
```diff
@@ -422,17 +448,17 @@ jobs:
           echo "github.event_name: ${{ github.event_name }}"
 
-      - name: install qt
+      - name: Install Qt
```
It's not right to change from lower case to capitalization.  Look around, all names are using lower cases.

Please fix the style.

yungyuc commented on .github/workflows/devbuild.yml:
```diff
@@ -230,6 +243,7 @@ jobs:
 
     env:
       QT_DEBUG_PLUGINS: 1
+      PIP_BREAK_SYSTEM_PACKAGES: 1 # disabling PEP668
```
I like the use of an environment variable to control installation behavior and the comment.

yungyuc reviewed (changes requested):
- [x] Fix line width
- [x] Fix action name styles (use lower case)

terrychan999 committed:
fix line width and letter case

yungyuc commented:
@terrychan999 If you are ready for the next review, please leave a global comment.

terrychan999 commented:
> @terrychan999 If you are ready for the next review, please leave a global comment.

Thanks for the reminder. Ready for review.

yungyuc reviewed (approved):
LGTM.  Thanks a lot.  Since the CI mac runs are unstable now, I will merge after it comes back to be stable.

yungyuc reviewed (changes requested):
@terrychan999 The CI fix #309 invalidated this PR.  Could you please help rebase to the latest master and resolve the conflict?

yungyuc commented:
@terrychan999 Can you take a look?

terrychan999 commented:
> @terrychan999 The CI fix #309 invalidated this PR. Could you please help rebase to the latest master and resolve the conflict?

Rebased. Ready for review, thanks.

yungyuc commented:
Thanks a lot, @terrychan999

yungyuc commented:
This PR fixed #284
